### ISSUE TITLE
report an error when specifying variance on `TypeVar` with constraints

### ISSUE THREAD

**Issue by DetachHead** [DetachHead/basedpyright#893]
```py
from typing import TypeVar

T = TypeVar("T", int, str, contravariant=True)
```

this should be an error because variance has no effect on constrained `TypeVar`s, because they aren't real generics, they can't be specialized to a subtype or supertype, so the variance is completely useless

the fact that there's no error on this in either mypy or pyright leads to user confusion because they misinterpret constraints as bounds, eg. https://stackoverflow.com/a/52997867

imo constraints are a stupid, barely used feature and all it does is introduce more unnecessary confusion to the concept of generics that many people already find difficult to understand. so the error message should clearly explain the situation / link to documentation about it

relevant comment: https://github.com/python/mypy/issues/11379#issuecomment-958676009

**Comment by jorenham** [comment]
Exactly; they're not type parameters, they're bidirectional type *mappings*.

**Comment by jorenham** [comment]
And for some reason, `TypeVar("T", A, B)` automatically and inevitably gets a secret extra constraint added to it: `Any`.

**Comment by DetachHead** [comment]
lol really? can you share an example of that?

**Comment by jorenham** [comment]
It's the most concise example, but it's a pattern I've seen several times in the wild, e.g. with `os.PathLike`.

https://basedpyright.com/?pyrightVersion=1.21.1&typeCheckingMode=all&reportExplicitAny=false&reportAny=false&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCCKcANFAArgxgDGYANgLABQLLN9AhgM7dQASPACoALVGlGcYASW4BRJDBEBTEAGUYIAPIgAQnBjLuAbVHihALigAKbprIAjA0YCUAXWuUw1OvRcWWKCCoAAEEcARVeEDggBNlYFgxdFtlemAXKABaAD4oM3RLKAA6UrZWZg4ePi0AdxRVLWAiOAK0AOZg0PCwSLw4GKD4xKVxVPTM3MJiK1Li8pZhqDRlGAB9bkhlNdH0U2SJK1t7KCdDbndrMHrVNbBgDa2dg6tBbjbJGXlFFXVNHX05325jckzyHw6XRAqwAriAUFBrg0QHcHpsINtdmhiliFsxoQA3ZSceg7RDKawrdbozEHax1ZFNFptawuNksIA

**Comment by jorenham** [comment]
> you'll probably be glad to hear about [this new rule](https://docs.basedpyright.com/dev/benefits-over-pyright/new-diagnostic-rules/#reportexplicitany) in the next release

In NumPy there are abstract *scalar types* such as `numpy.number`, and they often have an *in*variant type-parameter, but it's not really annotating anything, it's just labels the "n-bits" of scalar type. But because it's invariant, you can't do `np.floating[npt.NBitBase] = np.float32` (`npt.NBitBase` being the parameter's upper bound). So the only way to properly use it in practice, is as `np.floating[Any]`.
 
So to make a long story short; I think it would certainly be helpful to only have to supress the error when intentionally using `Any`, instead of having to do ignore it each time that some `Any`-ish type is referenced.

**Comment by DetachHead** [comment]
just to make sure we're on the same page, you mean you'd prefer to enable `reportExplicitAny` but not `reportAny`?

**Comment by jorenham** [comment]
> just to make sure we're on the same page, you mean you'd prefer to enable `reportExplicitAny` but not `reportAny`?

If I understand the docs correctly, then yes 👌🏻.

**Comment by KotlinIsland** [comment]
this is p1 imo

**Comment by jorenham** [comment]
> > just to make sure we're on the same page, you mean you'd prefer to enable `reportExplicitAny` but not `reportAny`?
> 
> If I understand the docs correctly, then yes 👌🏻.

Well, apparently I didn't: I expected the amount of `reportExplicitAny` errors to be less that those from `reportError`. 
But in `scipy-stubs` the former results in `3105` errors, and the latter in `1218` 🤔.

**Comment by DetachHead** [comment]
i guess [baselining](https://docs.basedpyright.com/latest/benefits-over-pyright/baseline/) those existing errors is an option if you still want to use this new rule
